File: algorithms/04_java/graphs/BipartiteCheck.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.List;

public class BipartiteCheck {
    public static boolean isBipartite(int n, List<List<Integer>> adj) {
        int[] color = new int[n];
        Arrays.fill(color, -1);
        for (int s = 0; s < n; s++) {
            if (color[s] != -1) continue;
            Deque<Integer> q = new ArrayDeque<>();
            color[s] = 0;
            q.add(s);
            while (!q.isEmpty()) {
                int u = q.poll();
                for (int v : adj.get(u)) {
                    if (color[v] == -1) {
                        color[v] = color[u] ^ 1;
                        q.add(v);
                    } else if (color[v] == color[u]) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    public static void main(String[] args) {
        int n = 4;
        List<List<Integer>> bip = new ArrayList<>();
        for (int i = 0; i < n; i++) bip.add(new ArrayList<>());
        bip.get(0).add(1); bip.get(0).add(3);
        bip.get(1).add(0); bip.get(1).add(2);
        bip.get(2).add(1); bip.get(2).add(3);
        bip.get(3).add(0); bip.get(3).add(2);
        if (!isBipartite(n, bip)) throw new AssertionError("graph should be bipartite");

        List<List<Integer>> notBip = new ArrayList<>();
        for (int i = 0; i < 3; i++) notBip.add(new ArrayList<>());
        notBip.get(0).add(1); notBip.get(1).add(0);
        notBip.get(1).add(2); notBip.get(2).add(1);
        notBip.get(2).add(0); notBip.get(0).add(2);
        if (isBipartite(3, notBip)) throw new AssertionError("triangle should not be bipartite");
        System.out.println("[Java BipartiteCheck] Bipartite check verified");
    }
}

File: algorithms/02_c/graphs/bipartite_check.c
#include <stdio.h>

#define MAXV 100

int is_bipartite(const int adj[][MAXV], int n) {
    int color[MAXV];
    for (int i = 0; i < n; ++i) color[i] = -1;

    for (int s = 0; s < n; ++s) {
        if (color[s] != -1) continue;
        color[s] = 0;
        int q[MAXV], front = 0, rear = 0;
        q[rear++] = s;
        while (front < rear) {
            int u = q[front++];
            for (int v = 0; v < n; ++v) {
                if (!adj[u][v]) continue;
                if (color[v] == -1) {
                    color[v] = color[u] ^ 1;
                    q[rear++] = v;
                } else if (color[v] == color[u]) {
                    return 0;
                }
            }
        }
    }
    return 1;
}

int main(void) {
    int even[MAXV][MAXV] = {{0}};   /* 4-cycle 0-1-2-3-0 */
    int tri[MAXV][MAXV] = {{0}};    /* triangle 0-1-2-0 */

    int ev[][2] = {{0, 1}, {1, 2}, {2, 3}, {3, 0}};
    int tr[][2] = {{0, 1}, {1, 2}, {2, 0}};
    for (int i = 0; i < 4; ++i) { even[ev[i][0]][ev[i][1]] = even[ev[i][1]][ev[i][0]] = 1; }
    for (int i = 0; i < 3; ++i) { tri[tr[i][0]][tr[i][1]] = tri[tr[i][1]][tr[i][0]] = 1; }

    if (!is_bipartite(even, 4)) {
        printf("[C Bipartite] FAILED: even cycle should be bipartite\n");
        return 1;
    }
    if (is_bipartite(tri, 3)) {
        printf("[C Bipartite] FAILED: triangle should not be bipartite\n");
        return 1;
    }
    printf("[C Bipartite] 2-colorability verified\n");
    return 0;
}
